### PR TITLE
DOC: Add missing imports, definitions and dummy file

### PR DESCRIPTION
**Pull request by pdebuyl**
Related to #14970 

Here, I fix the rst tests for `distutils.rst`:
- Add dummy file foo.c at runtime
- I added missing imports in an explicit manner, so as not to hide the information for readers

### CODE REVIEW

File: doc/source/reference/distutils.rst [hunk 1]
```diff
@@ -117,6 +120,8 @@ install the C library, you just use the method `add_installed_library` instead o
 `add_library`, which takes the same arguments except for an additional
 ``install_dir`` argument::
 
+  >>> import numpy.distutils.misc_util
+  >>> config = np.distutils.misc_util.Configuration(None, '', '.')
```
Comment (eric-wieser) [inline doc/source/reference/distutils.rst]: I think we either need to adjust the text to explain these lines, or hide them in a comment like the one at the top

Comment (pdebuyl) [reply doc/source/reference/distutils.rst]: I hesitated between the two. I can move both lines to the hidden block at the top.

File: doc/source/reference/distutils.rst [hunk 2]
```diff
@@ -1,5 +1,7 @@
 .. for doctests
    >>> with open('foo.c', 'w') as f: pass
+   >>> import numpy.distutils.misc_util
+   >>> config = np.distutils.misc_util.Configuration(None, '', '.')
```
Comment (eric-wieser) [inline doc/source/reference/distutils.rst]: Can you move this block down to right above the example that needs it? I don't think putting this at the top adds clarity.

**Comment by pdebuyl** [comment]
Should I squash the commits?

**Comment by eric-wieser** [comment]
Sure, saves me doing it.

Can you add a `DOC:` prefix to the commit message, and make clear in the message that the goal is to make the rst files pass refguide?

File: doc/source/reference/distutils.rst [hunk 3]
```diff
@@ -117,6 +120,11 @@ install the C library, you just use the method `add_installed_library` instead o
 `add_library`, which takes the same arguments except for an additional
 ``install_dir`` argument::
 
+.. for doctests
+
```
Comment (charris) [inline doc/source/reference/distutils.rst]: I think this would read easier without the blank lines.

Comment (pdebuyl) [reply doc/source/reference/distutils.rst]: sphinx complains about the version without a blank line, see failed circle ci

Comment (charris) [reply doc/source/reference/distutils.rst]: Interesting. I based the suggestion on other uses of `.. for doctests` in the documentation and it seems they may be in error.

Comment (charris) [reply doc/source/reference/distutils.rst]: Actually, looks like the indentation of one line is off.

**Comment by pdebuyl** [comment]
I can only get the "original" version to pass:

```
.. for doctests

   >>> import numpy.distutils.misc_util
   >>> config = np.distutils.misc_util.Configuration(None, '', '.')

  >>> config.add_installed_library('foo', sources=['foo.c'], install_dir='lib')
```

Should I revert to it?

File: doc/source/reference/distutils.rst [hunk 4]
```diff
@@ -117,6 +120,10 @@ install the C library, you just use the method `add_installed_library` instead o
 `add_library`, which takes the same arguments except for an additional
 ``install_dir`` argument::
 
+.. for doctests
+    >>> import numpy.distutils.misc_util
+    >>> config = np.distutils.misc_util.Configuration(None, '', '.')
+
   >>> config.add_installed_library('foo', sources=['foo.c'], install_dir='lib')
```
Comment (charris) [inline doc/source/reference/distutils.rst]: This line needs to be correctly indented.

**Comment by pdebuyl** [comment]
For info, the travis build succeeded (even though the CI badge is not updated on this PR page, you can simply click on it).

Also, I can squash the PR when/if necessary.

The current indent is

```
  .. for doctests
    >>> import numpy.distutils.misc_util
    >>> config = np.distutils.misc_util.Configuration(None, '', '.')

  >>> config.add_installed_library('foo', sources=['foo.c'], install_dir='lib')
```

(I indented the whole `..` block)

**Comment by charris** [comment]
OK, that works. But look where `.. for doctests` is located in

- `doc/source/reference/arrays.classes.rst`
- `doc/source/reference/maskedarray.baseclass.rst`

**Comment by pdebuyl** [comment]
Initially, the `for doctests` was at the top. I moved it on request https://github.com/numpy/numpy/pull/15616#pullrequestreview-361864421

I'll let you (i.e. NumPy maintainers) decide where it should ultimately go.

**Comment by charris** [comment]
@eric-wieser Ping.

**Comment by pdebuyl** [comment]
Hi @eric-wieser @charris 

The choice is between "import in comment at the top of the file" or "import in comment above line that requires it". I have no preference but this seems like a minor issue that could get merged without requiring too much of maintainer time, decreasing the PR count by 1 :-)

File: doc/source/reference/distutils.rst [hunk 5]
```diff
@@ -1,3 +1,6 @@
+.. for doctests
+   >>> with open('foo.c', 'w') as f: pass
+
```
Comment (eric-wieser) [inline doc/source/reference/distutils.rst]: Seems odd to leave this at the top while the others are at the bottom

**Comment by eric-wieser** [comment]
@charris: Putting the hidden doctest code right before it's needed makes it easier to move doc content around or delete it without leaving behind stale code at the top the file - so I prefer the approach taken by this PR.

File: doc/source/reference/distutils.rst (hunk 3 above)
Comment (eric-wieser) [reply doc/source/reference/distutils.rst]: How were you able to resolve the warning, @pdebuyl?

File: doc/source/reference/distutils.rst [hunk 6]
```diff
@@ -117,6 +120,10 @@ install the C library, you just use the method `add_installed_library` instead o
 `add_library`, which takes the same arguments except for an additional
 ``install_dir`` argument::
 
+  .. for doctests
```
Comment (eric-wieser) [inline doc/source/reference/distutils.rst]: Perhaps we should be more explicit here:
```suggestion
  .. hidden in a comment so as to be included in refguide but not rendered documentation
```

File: doc/source/reference/distutils.rst (hunk 5 above)
Comment (pdebuyl) [reply doc/source/reference/distutils.rst]: Fair enough, it will move :-)

File: doc/source/reference/distutils.rst (hunk 3 above)
Comment (pdebuyl) [reply doc/source/reference/distutils.rst]: By indenting the comment marker by two extra spaces.

File: doc/source/reference/distutils.rst (hunk 5 above)
Comment (pdebuyl) [reply doc/source/reference/distutils.rst]: It is even more useful that the import and the creation of the file foo.c are necessary at the same place.

**Review by eric-wieser — APPROVED** [review]
Looks good - can you check that this passes refguide still, assuming that CI does not yet check that?

**Comment by pdebuyl** [comment]
It passes the `refguide_check.py --rst` with current master. Should I squash now?

**Comment by mattip** [comment]
Thanks @pdebuyl, github can do the squash-merge.

**Comment by pdebuyl** [comment]
Thanks :-)

**Comment by eric-wieser** [comment]
@mattip: is the some way we can maintain a blacklist of broken rst files in git, and remove them one at a time alongside these PRs?

**Pull request merged**